include/unifex/any_ref.hpp

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/detail/vtable.hpp>
+#include <unifex/detail/with_type_erased_tag_invoke.hpp>
+#include <unifex/std_concepts.hpp>
+
+#include <utility>
+#include <type_traits>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+
+template <typename... CPOs>
+struct _any_ref {
+  class type;
+};
+
+template <typename... CPOs>
+class _any_ref<CPOs...>::type
+  : private with_type_erased_tag_invoke<type, CPOs>... {
+
+  // Inline up to 3 vtable-entries. Making a max size of 4 pointers.
+  using vtable_holder_t = std::conditional_t<
+    (sizeof...(CPOs) <= 3),
+    detail::inline_vtable_holder<CPOs...>,
+    detail::indirect_vtable_holder<CPOs...>>;
+
+ public:
+  template (typename Concrete)
+    (requires (!same_as<Concrete const, type const>))
+  /*implicit*/ type(Concrete& impl)
+    : vtable_(vtable_holder_t::template create<Concrete>())
+    , impl_((void*) std::addressof(impl)) {}
+
+  void swap(type& other) noexcept {
+    using std::swap;
+    swap(vtable_, other.vtable_);
+    swap(impl_, other.impl_);
+  }
+
+  // Two any_ref's compare equal IFF they refer to the same object (shallow).
+  //
+  // TODO: This is not entirely accurate as it might be possible to have two
+  // any_ref objects constructed from the same object at the same address but
+  // with different static-types (e.g. one any_ref is constructed from a base-class)
+  // and thus they would have different vtables. However, we can't just compare
+  // vtables since it's not guaranteed that function pointers will point to the
+  // same address.
```

Comment:
I guess it depends what we want "equality" to mean for type-erased types.
For other type-erasing wrappers it will generally check that the typeid is the same and that `operator==()` returns `true`.

With `any_ref` it's possible to construct two different views onto the same object that have vtables generated for different types and this might cause the `any_ref` objects to behave differently - they might dispatch to different static implementations.

Also, defining `any_ref` purely in terms of object identity means that it won't return `true` for cases where two different concrete objects of the same type that compare equal are wrapped in `any_ref` objects and you compare those `any_ref` objects.

e.g. do we want this to pass?
```
concrete a = ...;
concrete b = ...;
assert(a == b);
any_ref<foo_cpo> aRef = a;
any_ref<foo_cpo> bRef = b;
assert(aRef == bRef);
```

Reply:
Thanks for that elaboration; it makes the question clearer, and spurred a thought: perhaps there should be a well-defined, low-friction way to opt into a virtual implementation of equality.

If you provide it, you get deep equality, and if you don't, you get pointer equality.

I like the pay-for-what-you-use nature of this proposal, but I wonder what it costs in terms of complexity, or the evils of making everything configurable.